gobblin-compaction/src/test/java/org/apache/gobblin/compaction/mapreduce/orc/OrcKeyComparatorTest.java

```diff
@@ -0,0 +1,315 @@
+    Assert.assertEquals(conf.get(OrcConf.MAPRED_SHUFFLE_KEY_SCHEMA.getAttribute()), schema.toString());
+    comparator.setConf(conf);
+
+    // base record
+    OrcStruct record0 = (OrcStruct) OrcStruct.createValue(schema);
+    record0.setFieldValue("a", new IntWritable(1));
+    OrcList orcList0 = createOrcList(3, listSchema, 3);
+    record0.setFieldValue("b", orcList0);
+
+    // the same as base but different object, expecting equal to each other.
+    OrcStruct record1 = (OrcStruct) OrcStruct.createValue(schema);
+    record1.setFieldValue("a", new IntWritable(1));
+    OrcList orcList1 = createOrcList(3, listSchema, 3);
+    record1.setFieldValue("b", orcList1);
+
+    // Diff in int field
+    OrcStruct record2 = (OrcStruct) OrcStruct.createValue(schema);
+    record2.setFieldValue("a", new IntWritable(2));
+    OrcList orcList2 = createOrcList(3, listSchema, 3);
+    record2.setFieldValue("b", orcList2);
+
+    // Diff in array field: 1
+    OrcStruct record3 = (OrcStruct) OrcStruct.createValue(schema);
+    record3.setFieldValue("a", new IntWritable(1));
+    OrcList orcList3 = createOrcList(3, listSchema, 5);
+    record3.setFieldValue("b", orcList3);
+
+    // Diff in array field: 2
+    OrcStruct record4 = (OrcStruct) OrcStruct.createValue(schema);
+    record4.setFieldValue("a", new IntWritable(1));
+    OrcList orcList4 = createOrcList(4, listSchema, 3);
+    record4.setFieldValue("b", orcList4);
+
+    OrcKey orcKey0 = new OrcKey();
+    orcKey0.key = record0;
+    OrcKey orcKey1 = new OrcKey();
+    orcKey1.key = record1;
+    OrcKey orcKey2 = new OrcKey();
+    orcKey2.key = record2;
+    OrcKey orcKey3 = new OrcKey();
+    orcKey3.key = record3;
+    OrcKey orcKey4 = new OrcKey();
+    orcKey4.key = record4;
+
+    Assert.assertTrue(comparator.compare(orcKey0, orcKey1) == 0);
+    Assert.assertTrue(comparator.compare(orcKey1, orcKey2) < 0);
+    Assert.assertTrue(comparator.compare(orcKey1, orcKey3) < 0);
+    Assert.assertTrue(comparator.compare(orcKey1, orcKey4) < 0);
+  }
+
+  @Test
+  public void testComplexRecordMap() throws Exception {
+    OrcKeyComparator comparator = new OrcKeyComparator();
+    Configuration conf = new Configuration();
+    TypeDescription mapFieldSchema =
+        TypeDescription.createMap(TypeDescription.createString(), TypeDescription.createString());
+    TypeDescription schema =
+        TypeDescription.createStruct().addField("a", TypeDescription.createInt()).addField("b", mapFieldSchema);
+
+    conf.set(OrcConf.MAPRED_SHUFFLE_KEY_SCHEMA.getAttribute(), schema.toString());
+    Assert.assertEquals(conf.get(OrcConf.MAPRED_SHUFFLE_KEY_SCHEMA.getAttribute()), schema.toString());
+    comparator.setConf(conf);
+
+    // base record
+    OrcStruct record0 = (OrcStruct) OrcStruct.createValue(schema);
+    record0.setFieldValue("a", new IntWritable(1));
+    OrcMap orcMap = createOrcMap(new Text("key"), new Text("value"), mapFieldSchema);
```

Comment:
Should test that a two maps with the same set of entries in different order compares to be the same when the map contains more than 1 entry.

Reply:
Actually the internal implementation of `OrcMap` is a `Treemap`, which means we probably need to rule it out in comparison.